File: 2472-build-a-matrix-with-conditions/2472-build-a-matrix-with-conditions.java
class Solution {
    private int[] topo(int[][] edges, int n) {
        List<Integer>[] adj = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) adj[i] = new ArrayList<Integer>();

        int[] in_deg = new int[n + 1], vert = new int[n];

        for (int[] e: edges) {
            adj[e[0]].add(e[1]);
            in_deg[e[1]]++;
        }

        int idx = 0, t;
        Queue<Integer> q = new LinkedList<>();

        for (int i = 1; i <= n; i++) if (in_deg[i] == 0) q.offer(i);

        while (!q.isEmpty()) {
            t = q.poll();
            vert[idx++] = t;
            n--;
            for (int v: adj[t]) if (--in_deg[v] == 0) q.offer(v);
        }

        if (n != 0) return new int[0];

        return vert;
    }
    public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
        int[] row = topo(rowConditions, k);
        int[] col = topo(colConditions, k);

        if (row.length == 0 || col.length == 0) return new int[0][0];

        int[][] ans = new int[k][k];
        for (int r = 0; r < k; r++) 
            for (int c = 0; c < k; c++) 
                if (row[r] == col[c]) ans[r][c] = row[r];

        return ans;
    }
}

File: 2472-build-a-matrix-with-conditions/2472-build-a-matrix-with-conditions.cpp
class Solution {
private:
    vector<int> topo(vector<vector<int>> edges, int n) {
        vector<int>* adj = new vector<int>[n + 1]; 

        vector<int> in_deg(n + 1, 0), vert;

        for (vector<int>& e: edges) {
            adj[e[0]].push_back(e[1]);
            in_deg[e[1]]++;
        }

        int t;
        queue<int> q;

        for (int i = 1; i <= n; i++) if (in_deg[i] == 0) q.push(i);

        while (!q.empty()) {
            t = q.front();
            vert.push_back(t);
            q.pop();
            n--;
            for (int v: adj[t]) if (--in_deg[v] == 0) q.push(v);
        }

        if (n) return {};

        return vert;
    }
public:
    vector<vector<int>> buildMatrix(int k, vector<vector<int>>& rowConditions, vector<vector<int>>& colConditions) {
        vector<int> row = topo(rowConditions, k);
        vector<int> col = topo(colConditions, k);

        if (row.empty() || col.empty()) return {};

        vector<vector<int>> ans(k, vector<int>(k, 0));
        for (int r = 0; r < k; r++) 
            for (int c = 0; c < k; c++) 
                if (row[r] == col[c]) ans[r][c] = row[r];

        return ans;
    }
};